Feature/snap yaml

mvo5 opened the pull request:
This branch uses meta/snap.yaml instead of package.yaml.

To make the transition easier it also uses the new "skills" system and provides a "migration-skill" that takes exactly the same arguments as the previous security system.

mvo5 commented:
Fwiw, the integration tests will fail because we have no snaps in the store that are converted to the new meta/snap.yaml. But once there is a chance that this lands I can prepare updates.

chipaca commented on docs/frameworks.md:
```diff
@@ -147,7 +147,7 @@ might contain something like:
 ### App yaml
 
-For apps wanting to use a particular framework, meta/package.yaml simply
+For apps wanting to use a particular framework, meta/snap.yaml simply
```
while you're here, can you backtick-quote `meta/snap.yaml`?

chipaca commented on docs/meta.md:
```diff
@@ -21,8 +15,9 @@ for a snap package. The following keys are mandatory:
-* `icon`: a SVG icon for the snap that is displayed in the store
-* `explicit-license-agreement`: set to `Y` if the user needs to accept a
+* `summary`: A short summary
+* `description`: A long description
```
before the description was optional (and you got a placeholder if it wasn't there). Is it now required?

mvo5 replied:
I listed it under "optional" in the docs for now.  The new "spec" does not actually specify what is optional and what is required yet. I'm happy to tweak as needed.

chipaca commented on docs/meta.md:
```diff
@@ -21,8 +15,9 @@ for a snap package. The following keys are mandatory:
+* `license-agreement`: set to `explicit` if the user needs to accept a
   special `meta/license.txt` before the snap can be installed
 * `license-version`: a string that, when it changes and
   `explicit-license-agreement` is `Y`, prompts the user to accept the
```
This needs to change to “and `license-agreement` is set to `explicit`”

mvo5 replied:
Good catch, thanks!

chipaca commented on snappy/errors.go:
```diff
@@ -92,7 +88,7 @@ var (
-	// ErrInvalidFrameworkSpecInYaml is returned if a package.yaml
+	// ErrInvalidFrameworkSpecInYaml is returned if a snap.yaml
 	// has both frameworks and framework entries.
 	ErrInvalidFrameworkSpecInYaml = errors.New("yaml can't have both \"frameworks\" and (deprecated) \"framework\" keys")
```
are we still using this error? (why? where?)

chipaca commented on snappy/snap_yaml.go:
```diff
@@ -79,72 +86,50 @@ type ServiceYaml struct {
-
-	return nil
-}
-
```
Hurray for killing `deprecarch`

chipaca commented:
Big :+1:. A few little things.

mvo5 commented:
Thanks a bunch for the review. I fixed the outlined issues now.

zyga commented on snappy/click.go:
```diff
@@ -44,27 +44,39 @@ import (
+
+	if uses.Type != "migration-skill" {
+		return fmt.Errorf("can not use skill %q, only migration-skill supported", uses.Type)
+	}
```
In the world where we do know about skills, where would the knowledge of known skill types (and valid values) come from?

niemeyer replied:
There's no need to. The migration-skill is an aberration we're putting in place to move people forward comfortably. We don't need a complete list here.

zyga commented on snappy/click.go:
```diff
@@ -316,12 +326,12 @@ func (m *packageYaml) addPackageServices(baseDir string, inhibitHooks bool, inte
-		if m.Type == snap.TypeFramework && service.BusName != "" {
-			content, err := genBusPolicyFile(service.BusName)
+		if m.Type == snap.TypeFramework && app.BusName != "" {
+			content, err := genBusPolicyFile(app.BusName)
```
This is something that skills will interact with heavily. How can that method know about skills.Repository()?

This is not a blocker in any way, reading this makes me think about integration of skills into the overlord.

niemeyer replied:
The organization here needs much improvement. The structure that holds metadata for a package  (this is a _snapYaml!_) should never be dealing with security by itself.

mvo5 replied:
Indeed, sorry for that. Is https://github.com/mvo5/snappy/compare/feature/snap-yaml3...mvo5:feature/snap-yaml-refactor-type?expand=1 a start in the right direction?

zyga commented on snappy/click.go:
```diff
@@ -411,13 +426,17 @@ func (m *packageYaml) removePackageServices(baseDir string, inter interacter) er
+			continue
+		}
+
+		aaProfile, err := getSecurityProfile(m, app.Name, baseDir)
```
More intersection with skills (again, this is not a blocking comment).

How do we use skills to generate _all_ security files? How do we marry this with the state engine and transactionality?

niemeyer replied:
Again, this is misplaced. Don't worry about getting skills here.

The question is also misplaced, by the way.

zyga commented on snappy/snap_local.go:
```diff
@@ -174,11 +168,12 @@ func (s *SnapPart) Channel() string {
 	}
 
-	return filepath.Join(s.basedir, s.m.Icon)
+	return found[0]
```
Hmm, this might return various different icons depending on unpredictable filesystem ordering. I'd much rather sort the list of found strings here and return a stable result (whatever it is).

chipaca replied:
Glob sorts

zyga replied:
Ah, thanks, then this is +1

niemeyer commented on daemon/api.go:
```diff
@@ -342,7 +342,7 @@ var findServices = snappy.FindServices
 
 type svcDesc struct {
```
appDesc?

niemeyer commented on daemon/api.go:
```diff
@@ -408,7 +408,7 @@ func snapService(c *Command, r *http.Request) Response {
 		return InternalError("active snap is not a *snappy.SnapPart: %T", ipart)
 	}
-	svcs := part.ServiceYamls()
+	svcs := part.Apps()
```
apps?

niemeyer commented on daemon/api.go:
```diff
@@ -416,7 +416,10 @@ func snapService(c *Command, r *http.Request) Response {
 
 	svcmap := make(map[string]*svcDesc, len(svcs))
```
appmap

niemeyer commented on daemon/api.go:
```diff
@@ -416,7 +416,10 @@ func snapService(c *Command, r *http.Request) Response {
+		svcmap[svcs[i].Name] = &svcDesc{Spec: svcs[i], Op: action}
 	}
 
 	if svcName != "" && svcmap[svcName] == nil {
```
appName, etc.. not sure about what else in the file needs updating, but worth a look.

niemeyer commented on daemon/snappy_part_iface_test.go:
```diff
@@ -53,7 +53,7 @@ type tP struct {
 	frameworksErr error
 
-	svcYamls []snappy.ServiceYaml
+	svcYamls map[string]*snappy.AppYaml
```
Same here.

niemeyer commented on integration-tests/testutils/store/store_test.go:
```diff
@@ -140,12 +140,8 @@ func (s *storeTestSuite) makeTestSnap(c *C, packageYamlContent string) string {
-
-	content := "Random\nExample"
-	err := ioutil.WriteFile(readmeMd, []byte(content), 0644)
+	packageYaml := filepath.Join(tmpdir, "meta", "snap.yaml")
```
snapYaml

niemeyer commented on snap/squashfs/squashfs.go:
```diff
@@ -127,7 +127,7 @@ const (
 // Info returns information like name, type etc about the package
 func (s *Snap) Info() (*snap.Info, error) {
-	packageYaml, err := s.ReadFile("meta/package.yaml")
+	packageYaml, err := s.ReadFile("meta/snap.yaml")
```
snap

niemeyer commented on snappy/click.go:
```diff
@@ -44,27 +44,39 @@ import (
-func verifyBinariesYaml(binary Binary) error {
-	return verifyStructStringsAgainstWhitelist(binary, servicesBinariesStringsWhitelist)
+func verifyAppYaml(app *AppYaml) error {
+	return verifyStructStringsAgainstWhitelist(*app, servicesBinariesStringsWhitelist)
```
Why the dereferencing here an below (`*app` and `*uses`)?

niemeyer commented on snappy/click.go:
```diff
@@ -154,7 +166,6 @@ ubuntu-core-launcher {{.UdevAppName}} {{.AaProfile}} {{.Target}} "$@"
 // verifyStructStringsAgainstWhitelist takes a struct and ensures that
 // the given whitelist regexp matches all string fields of the struct
 func verifyStructStringsAgainstWhitelist(s interface{}, whitelist *regexp.Regexp) error {
```
The magic is deep on this one. :-)  Seems like explicitly validating the few fields we care about would be straightforward and leave less legroom for surprises. But I guess it doesn't matter much now, since click.go should be going away soon anyway.

mvo5 replied:
Thanks, I added a FIXME. I agree this is a overthetop.

niemeyer commented on snappy/errors.go:
```diff
@@ -220,7 +212,7 @@ type ErrStructIllegalContent struct {
 
 func (e *ErrStructIllegalContent) Error() string {
-	return fmt.Sprintf("services description field '%s' contains illegal '%s' (legal: '%s')", e.Field, e.Content, e.Whitelist)
+	return fmt.Sprintf("services description field '%s' contains illegal %q (legal: '%s')", e.Field, e.Content, e.Whitelist)
```
s/services/application/?

niemeyer commented on snappy/parts.go:
```diff
@@ -35,9 +35,9 @@ import (
-// ServiceYamler implements snappy packages that offer services
-type ServiceYamler interface {
-	ServiceYamls() []ServiceYaml
+// AppYamler implements snappy packages that offer services
```
I'm not sure if I'm more confused with or without the comment. :-)

s/snappy packages/snaps/
s/services/applications/

But I still don't know what it means. It's also somewhat surprising to see all of that part of the public interface. Perhaps that's all part of the fat that is going away with the single snap types you're working on?

niemeyer commented on snappy/snap_yaml.go:
```diff
@@ -54,17 +54,24 @@ type Ports struct {
+	Version  string   `yaml:"version"`
+	Command  string   `yaml:"command"`
+	Provides []string `yaml:"provides"`
+	Consumes []string `yaml:"consumes"`
```
These don't look right. Should be `"offers"` & `"uses"`.

niemeyer commented:
This is a massive change, but one that looks relatively innocent in the sense of it being a fixup of existing logic mostly. There are a few more instances of services to be correct to apps around the code, but most of my anxiety while reading the code is related to pieces unrelated to the changes being done. I hadn't realized how much non-metadata logic we have attached to these metadata files, and some of it is even exposed. More things for our TODO list.

mvo5 commented:
@niemeyer Thanks for this excellent review. I think I addressed the points raised now.

Sorry about the need for more TODOs. I expect a lot of the current security and file creation bits to move into the overlord(?). If that is correct and the first overlord bit lands, I'm happy to work on further refactor and I'm very happy to get input into what architecture you prefer most here.

Thanks again,
 Michael

sergiusens commented on snappy/click.go:
```diff
@@ -190,87 +203,83 @@ func verifyStructStringsAgainstWhitelist(s interface{}, whitelist *regexp.Regexp
-			Forking:        service.Forking,
+			IsNetworked:    app.Ports != nil && len(app.Ports.External) > 0,
+			BusName:        app.BusName,
+			Forking:        app.Forking,
```
Why not just implement `app.Daemon` which should hold valid `systemd` values (whitelisted `simple`, `forking`, ...) and use that as a replacement string?

sergiusens commented on daemon/api.go:
```diff
@@ -408,48 +408,51 @@ func snapService(c *Command, r *http.Request) Response {
-		svcmap[svcs[i].Name] = &svcDesc{Spec: &svcs[i], Op: action}
+	appmap := make(map[string]*appDesc, len(apps))
+	for i := range apps {
+		if apps[i].Daemon == "" {
```
is there any place where we are checking for valid `app[i].Daemon` values?

mvo5 replied:
Nice catch, thanks! It was missing but now its there.

sergiusens commented on daemon/api_test.go:
```diff
@@ -923,10 +930,10 @@ func (s *apiSuite) TestServiceLogs(c *check.C) {
-	// have an icon for it in the snap itself
-	iconfile := filepath.Join(dirs.SnapSnapsDir, "foo.bar", "v1", "icon.ick")
+	// have an icon for it in the package itself
+	iconfile := filepath.Join(dirs.SnapSnapsDir, "foo.bar", "v1", "meta", "icon.ick")
```
I thought we were just supporting svg and png, how is `ick` picked up?

sergiusens commented:
I did a quick pass, looks generally good.

niemeyer commented:
LGTM too, thanks Michael!
